gym_adserver/envs/adserver.py

```diff
@@ -11,38 +11,111 @@
+
+AD_TYPES = ['car', 'food', 'electronics', 'clothing']
 
 class AdServerEnv(gym.Env):
     metadata = {
         'render.modes': ['human']
     }
 
-    def __init__(self, num_ads, time_series_frequency, reward_policy=None):        
+    def __init__(self, num_ads, time_series_frequency, users=None, reward_policy=None, budgets=None, seed=None):        
         self.time_series_frequency = time_series_frequency        
         self.num_ads = num_ads
         self.reward_policy = reward_policy
-        self.click_probabilities = None
+        self.max_impressions = 300
+        self.seed_value = seed
+        self.seed(seed=self.seed_value)
+        self.ads_random = np.random.default_rng(seed)
+        
+        self.users = [User(['car', 'electronics'], np_random=self.ads_random),
+                      User(['food', 'clothing'], np_random=self.ads_random)]
+        
+        self.click_probabilities = [None] * num_ads
+        self.bids = [None] * num_ads
+        self.budgets = [None] * num_ads
+        self.ads = [None] * num_ads
+        self.revenues = [None] * num_ads
+        self.types = [None] * num_ads
+        
 
         # Initial state (can be reset later)
-        ads = [Ad(i) for i in range(num_ads)]
+        self._generate_ads()
         clicks = 0
         impressions = 0
-        self.state = (ads, impressions, clicks)
+        self.state = (self.ads, impressions, clicks)
         self.ctr_time_series = []
+        self.revenue_time_series = []
 
         # Environment OpenAI metadata
         self.reward_range = (0, 1)
         self.action_space = spaces.Discrete(num_ads) # index of the selected ad
         self.observation_space = spaces.Box(low=0.0, high=np.inf, shape=(2, num_ads), dtype=np.float) # clicks and impressions, for each ad
-
+            
+    def _generate_ads(self):
+        for i in range(self.num_ads):
+            ad_type = self.ads_random.choice(AD_TYPES)
+            click_probability = self.get_ad_type_click_probability(ad_type, 0)
+            bid = self._generate_bid()
+            budget = self._generate_budget()
+            ad = Ad(id=i, click_probability=click_probability, bid=bid, budget=budget, type=ad_type)
+            
+            self.ads[i] = ad
+            self.bids[i] = bid
+            self.click_probabilities[i] = click_probability
+            self.budgets[i] = budget
+            self.revenues[i] = ad.revenue
+            self.types[i] = ad.type
+
+    def _generate_bid(self):
+        # Your custom logic to generate bid based on ad_type
+        return self.ads_random.uniform(0.01, 1)
+
+    def _generate_budget(self):
+        # Your custom logic to generate budget based on ad_type
+        return self.ads_random.uniform(500, 1000)
+
+    def get_current_time_of_day(self):
+        _, impressions, _ = self.state
+        time_of_day = impressions / self.max_impressions
+        return time_of_day
+
+    def get_ad_type_click_probability(self, ad_type, time_of_day):
```

Comment:
I'm wondering if this can be simpler (I mean more readable). Anyway, can you please comment this method explaining the logic?